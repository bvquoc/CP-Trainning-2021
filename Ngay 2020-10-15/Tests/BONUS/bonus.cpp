#include <bits/stdc++.h>
using namespace std;

#define a(i, j) aa[(i - 1) * n + j]
#define sum(i, j) sumsum[(i - 1) * n + j]

long long sumsum[1010101];
long long aa[1010101];

int main() {
    freopen("bonus.inp", "r", stdin);
    freopen("bonus.out", "w", stdout);
    ios_base::sync_with_stdio(false);

    int m, n, q;
    cin >> m >> n >> q;
    for (int i = 1; i <= m; i++)
        for (int j = 1; j <= n; j++)
            cin >> a(i, j);

    for (int i = 1; i <= m; i++)
        for (int j = 1; j <= n; j++)
            sum(i, j) = sum(i - 1, j) + sum(i, j - 1) - sum(i - 1, j - 1) + a(i, j);

    while (q--) {
        int u, v, x, y;
        cin >> u >> v >> x >> y;
        cout << sum(x, y) - sum(u - 1, y) - sum(x, v - 1) + sum(u - 1, v - 1) << endl;
    }

    return 0;
}
