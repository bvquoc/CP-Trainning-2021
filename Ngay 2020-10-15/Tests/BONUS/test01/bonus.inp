3 15 45
8 4 8 3 6 2 1 1 4 6 5 6 1 10 7
4 10 1 4 4 10 5 3 8 10 2 6 2 1 1
8 7 2 4 4 3 2 2 3 9 4 7 10 1 1
3 1 3 6
1 1 2 3
3 6 3 10
1 3 2 12
1 2 1 14
2 7 3 9
1 4 2 12
1 1 2 2
2 4 3 6
1 10 1 14
2 13 2 14
1 5 2 15
1 3 2 6
1 12 2 15
1 6 2 7
1 6 2 13
2 4 3 8
2 3 3 11
3 1 3 5
3 3 3 13
1 4 3 6
1 1 3 12
1 11 3 13
3 4 3 4
3 4 3 5
1 8 2 14
1 1 2 7
1 2 1 14
2 5 3 14
1 2 3 10
2 8 3 9
1 1 3 4
1 6 3 10
2 1 3 9
2 11 3 11
2 2 2 8
1 2 3 9
1 1 3 2
1 10 3 12
2 4 3 15
1 2 2 4
2 2 3 12
2 3 3 10
1 5 3 6
1 9 2 14

