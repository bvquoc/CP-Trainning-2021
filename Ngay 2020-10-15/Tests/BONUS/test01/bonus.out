28
35
19
95
57
23
86
26
29
28
3
101
38
34
18
72
41
80
25
50
40
176
43
4
8
65
70
57
96
126
16
63
69
84
6
37
101
41
55
106
30
110
74
29
61
